pkg/kubediscovery/handlers.go

```diff
@@ -63,75 +41,137 @@ type caLoader interface {
+		if err != nil {
+			return "", err
+		}
+		if err := json.Unmarshal(data, &tl.tokenMap); err != nil {
+			return "", err
+		}
 	}
+	if val, ok := tl.tokenMap[tokenID]; ok {
+		return val, nil
+	}
+	return "", errors.New(fmt.Sprintf("invalid token: %s", tokenID))
+}
 
-	return string(data), nil
+type endpointsLoader interface {
+	LoadList() ([]string, error)
+}
+
+type jsonFileEnpointsLoader struct {
+	endpoints []string
+}
+
+func (el *jsonFileEnpointsLoader) LoadList() ([]string, error) {
+	if len(el.endpoints) == 0 {
+		data, err := ioutil.ReadFile(EndpointListPath)
+		if err != nil {
+			return nil, err
+		}
+		if err := json.Unmarshal(data, &el.endpoints); err != nil {
+			return nil, err
+		}
+	}
+	return el.endpoints, nil
 }
 
 // ClusterInfoHandler implements the http.ServeHTTP method and allows us to
 // mock out portions of the request handler in tests.
 type ClusterInfoHandler struct {
-	tokenLoader tokenLoader
-	caLoader    caLoader
+	tokenLoader     tokenLoader
+	caLoader        caLoader
+	endpointsLoader endpointsLoader
 }
 
 func NewClusterInfoHandler() *ClusterInfoHandler {
-	tl := hardcodedTokenLoader{}
-	cl := fsCALoader{}
 	return &ClusterInfoHandler{
-		tokenLoader: &tl,
-		caLoader:    &cl,
+		tokenLoader:     &jsonFileTokenLoader{},
+		caLoader:        &fsCALoader{},
+		endpointsLoader: &jsonFileEnpointsLoader{},
 	}
 }
 
 func (cih *ClusterInfoHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
-	tokenId := req.FormValue("token-id")
-	log.Printf("Got token ID: %s", tokenId)
-	token, err := cih.tokenLoader.Lookup(tokenId)
+	tokenID := req.FormValue("token-id")
+	log.Printf("Got token ID: %s", tokenID)
+	token, err := cih.tokenLoader.LoadAndLookup(tokenID)
 	if err != nil {
 		log.Printf("Invalid token: %s", err)
 		http.Error(resp, "Forbidden", http.StatusForbidden)
 		return
 	}
 	log.Printf("Loaded token: %s", token)
 
+	// TODO probably should not leak server-side errors to the client
 	caPEM, err := cih.caLoader.LoadPEM()
-	caB64 := base64.StdEncoding.EncodeToString([]byte(caPEM))
+	if err != nil {
+		err = fmt.Errorf("Error loading root CA certificate data: %s", err)
+		log.Println(err)
+		http.Error(resp, err.Error(), http.StatusInternalServerError)
+		return
+	}
 
+	endpoints, err := cih.endpointsLoader.LoadList()
 	if err != nil {
-		http.Error(resp, "Error encoding CA", http.StatusInternalServerError)
+		err = fmt.Errorf("Error loading list of API endpoints: %s", err)
+		log.Println(err)
+		http.Error(resp, err.Error(), http.StatusInternalServerError)
 		return
 	}
 
 	clusterInfo := ClusterInfo{
-		Type:             "ClusterInfo",
-		Version:          "v1",
-		RootCertificates: caB64,
+		CertificateAuthorities: []string{caPEM},
+		Endpoints:              endpoints,
 	}
 
 	// Instantiate an signer using HMAC-SHA256.
-	hmacTestKey := fromHexBytes(token)
-	signer, err := jose.NewSigner(jose.HS256, hmacTestKey)
+	signer, err := jose.NewSigner(jose.HS256, []byte(token))
```

Comment:
I need to investigate more but there might be a problem with this change, decoding the token (a string of hex characters) was from an example I was following in the JWS source itself and I think may be the correct behavior if you're using hex strings as an encryption key. https://github.com/square/go-jose/blob/master/jws_test.go#L267